File: lc_soln/0347_top_k_frequent_elements/solution.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        std::unordered_map<int, int> counter;
        for (int num : nums) {
            counter[num]++;
        }
        std::vector<std::vector<int>> buckets(nums.size() + 1, std::vector<int>());
        for (const auto& [num, freq] : counter) {
            buckets[freq].push_back(num);
        }
        std::vector<int> result;
        result.reserve(k);
        for (int i = buckets.size() - 1; i >= 0; --i) {
            for (int num : buckets[i]) {
                result.push_back(num);
                if (result.size() == k) {
                    return result;
                }
            }
        }
        return result;
    }
};

int main() {
    return 0;
}

File: lc_soln/0347_top_k_frequent_elements/solution.rs
struct Solution;

impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let k = k as usize;
        let n = nums.len();
        let mut counter: HashMap<i32, usize> = HashMap::with_capacity(n);
        for num in nums {
            *counter.entry(num).or_default(1) += 1;
        }
        let mut buckets: Vec<Vec<i32>> = vec![Vec::new(); n + 1];
        for (num, freq) in counter {
            buckets[freq].push(num);
        }
        let mut result = Vec::with_capacity(k);
        for bucket in buckets.into_iter().rev() {
            for num in bucket {
                result.push(num);
                if result.len() == k {
                    return result;
                }
            }
        }
        result
    }
}

fn main() {}
